nitrocli/src/main.rs

```diff
@@ -68,176 +68,31 @@
-    vu = get_volume_status(&status.unencrypted_volume),
-    ve = get_volume_status(&status.encrypted_volume),
-    vh = get_volume_status(&status.hidden_volume),
-  );
-}
-
-/// Inquire the status of the nitrokey.
-fn status() -> Result<()> {
-  let status = get_storage_device()?
-    .get_status()
-    .map_err(|err| get_error("Getting Storage status failed", &err))?;
-
-  print_status(&status);
-  Ok(())
-}
-
-/// Open the encrypted volume on the nitrokey.
-fn open() -> Result<()> {
-  let device = get_storage_device()?;
-
-  let mut retry = 3;
-  let mut error_msg: Option<&str> = None;
-  loop {
-    // TODO: Rethink the usage of String::from_utf8_lossy here. We may
-    //       not want to silently modify the password!
-    let passphrase = pinentry::inquire_passphrase(PIN_TYPE, error_msg)?;
-    let passphrase = String::from_utf8_lossy(&passphrase);
-    match device.enable_encrypted_volume(&passphrase) {
-      Ok(()) => return Ok(()),
-      Err(err) => match err {
-        nitrokey::CommandError::WrongPassword => {
-          pinentry::clear_passphrase(PIN_TYPE)?;
-          retry -= 1;
-
-          if retry > 0 {
-            error_msg = Some("Wrong password, please reenter");
-            continue;
-          }
-          let error = "Opening encrypted volume failed: Wrong password";
-          return Err(Error::Error(error.to_string()));
-        }
-        err => return Err(get_error("Opening encrypted volume failed", &err)),
-      },
-    };
-  }
-}
-
-#[link(name = "c")]
-extern "C" {
-  fn sync();
-}
-
-/// Close the previously opened encrypted volume.
-fn close() -> Result<()> {
-  // Flush all filesystem caches to disk. We are mostly interested in
-  // making sure that the encrypted volume on the nitrokey we are
-  // about to close is not closed while not all data was written to
-  // it.
-  unsafe { sync() };
-
-  get_storage_device()?
-    .disable_encrypted_volume()
-    .map_err(|err| get_error("Closing encrypted volume failed", &err))
-}
-
-/// Clear the PIN stored when opening the nitrokey's encrypted volume.
-fn clear() -> Result<()> {
-  pinentry::clear_passphrase(PIN_TYPE)
-}
-
-// A macro for generating a match of the different supported commands.
-// Each supplied command is converted into a string and matched against.
-macro_rules! commands {
-  ( $str:expr, [ $( $command:expr), *] ) => {
-    match &*$str.to_string() {
-      $(
-        stringify!($command) => {
-          if let Err(err) = $command() {
-            println!("{}", err);
-            return 1
-          }
-          return 0
-        },
-      )*
-      x => {
-        println!("Invalid command: {}", x);
-        println!("Available commands: {}", stringify!( $($command)* ));
-        return 1
-      },
-    }
-  }
-}
-
 fn run() -> i32 {
-  let argv: Vec<String> = std::env::args().collect();
-  if argv.len() != 2 {
-    println!("Usage: {} <command>", argv[0]);
-    return 1;
+  let (command, args) = options::parse_arguments();
```

Comment:
Instead of returning the command and `execute`ing it here, do you think it would be clearer to just short-circuit that, rename `parse_arguments` to `handle_arguments` and have it deal with the dispatch of the command?